### PR TITLE
Bluetooth: Mesh: split prov.c into two separate modules

### PR DESCRIPTION
**Pull request by LingaoM**
The previous version coupled `provisioner` and `device` provisioning in a source file `prov.c`.
For only supporting device or provisioner, this will not be compiled separately, and coupling
in a file will cause unclear logic, which is not conducive to later maintenance, for example, support for remote provisioning.

Now it is divided into three files and managed separately, which is called **prov.c** **prov_device.c** and **provisioner.c**.

Add `CONFIG_BT_MESH_PROV_DEVICE` in Kconfig  to control whether device provisioning is supported.
Use  `CONFIG_BT_MESH_PROVISIONER` in Kconfig  to control whether provisioner role is supported.

**Add support for all provisioner OOB authentication implementations.**
-  Support OOB Public Key
- Support Input OOB Authentication
- Support Ouput OOB Authentication
- Support Static OOB Authentication

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

### CODE REVIEW

**Comment by trond-snekvik** [comment]
cc @tsvehagen

File: subsys/bluetooth/mesh/provisioner.c
Comment (tsvehagen) [inline subsys/bluetooth/mesh/provisioner.c:55]: Since the file `provisioner.c` is only build when `CONFIG_BT_MESH_PROVISIONER` is set, this whole "trick" is not needed. I would suggest just doing

```
struct provisioner_link {
	struct bt_mesh_cdb_node *node;
	uint16_t addr;
	uint16_t net_idx;
	uint8_t attention_duration;
	uint8_t uuid[16];
} provisioner;
```

**Review by tsvehagen — CHANGES_REQUESTED** [review]
This is a very welcome change I think. Nice work.

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:55]: Great

**Comment by trond-snekvik** [comment]
Marking as draft, as the diff keeps changing. Please take it out of draft state when it's ready for review.

**Comment by LingaoM** [comment]
Remove draft since conflict has been solved and ready for review.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:601]: This is unused.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:139]: This should be determined by whether the application has called `bt_mesh_set_oob_public_key`. Ie, when we send start, check `REMOTE_PUB_KEY`, and request a public key if we haven't gotten one OOB. Just because a device has an OOB public key available, doesn't mean the provisioner has to follow the oob pubkey procedure.

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:285]:
```suggestion
int bt_mesh_prov_remote_pub_key_set(const uint8_t public_key[64]);
```

Also, need to document that this is a provisioner only API.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
All in all, this looks very good, but I have some comments on the auth_start mechanism and the layer split:

File: subsys/bluetooth/mesh/Kconfig
Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig:29]: If I remember correctly, `BT_MESH_PROV` (which indicates that we want support for the generic provisioning protocol) has so far been selected indirectly by choosing which provisioning bearer you want. Saying you want `BT_MESH_PROVISIONEE` doesn't tell you anything about provisioning bearers, so I'm not sure this is the right thing. So unless we want to break the semantics on the way that you choose mesh provisioning Kconfig options it seems to me like this `BT_MESH_PROVISIONEE` should `depend on BT_MESH_PROV` instead of selecting it.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/Kconfig
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/Kconfig:29]: Sounds good to me

File: include/bluetooth/mesh/main.h
Comment (LingaoM) [reply include/bluetooth/mesh/main.h:55]: int bt_mesh_auth_method_set_static(void); ??

File: subsys/bluetooth/mesh/provisioner.c
Comment (tsvehagen) [inline subsys/bluetooth/mesh/provisioner.c:51]: I think this should be renamed to something like `provisioning_state`. Simply `provisioner` is quite vague... (I'm to blame for that :p)

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h]: For static OOB authentication

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:676]: If the length exceeds 8, it will still fail in `prov_check_method`, because, it was checked whether the length is locally supported.

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:221]: When NO OOB is used, prov_device.size has been assigned a value of zero. I'm sorry that this is really obscure.

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:164]: Should put the key behind? https://github.com/zephyrproject-rtos/zephyr/blob/dc37f988e03029d43fb3c92d4f2f966e0c90e435/subsys/bluetooth/mesh/prov.c#L411-L416

File: subsys/bluetooth/mesh/prov_bearer.h
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov_bearer.h:127]: sure, I do want to put it in prov.h, but the `prov_bearer_send_complete_t ` function is defined in prov_bearer.h, If I put the send function in prov.h, then any source file containing prov.h will have to include prov_bearer.h.

File: subsys/bluetooth/mesh/prov.h
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov.h:65]: This length field currently has a copy in provisioner.c and provisionee.c. You mean to extract this length field separately, and only one is in prov.c？

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/prov.h]: Yes, creating a `static const uint16_t msg_lengths[10]` in prov_recv

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/provisioner.c:157]: Sorry, I got the endianness the wrong way, it should look like the prov_auth code, yes. I'll update the comment

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:317]:
```suggestion
/** @brief Use static OOB authentication.
 *
 *  Provisioner only.
 *  
 *  Instruct the unprovisioned device to use static OOB authentication, and use 
 *  the given static authentication value when provisioning.
 *
 *  @param static_val Static OOB value.
 *  @param size Static OOB value size.
 *
 *  @return Zero on success or (negative) error code otherwise.
 */
int bt_mesh_auth_method_set_static(const uint8_t *static_val, uint8_t size);
```

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:329]:
```suggestion
/** @brief Don't use OOB authentication.
 *
 *  Provisioner only.
 *  
 *  Don't use any authentication when provisioning new devices. This is the 
 *  default behavior.
 *  
 *  @warning Not using any authentication exposes the mesh network to 
 *           impersonation attacks, where attackers can pretend to be the 
 *           unprovisioned device to gain access to the network. Authentication
 *           is strongly encouraged.
 *           
 *  @return Zero on success or (negative) error code otherwise.
 */
int bt_mesh_auth_method_set_none(void);
```

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:185]: Still unresolved: Only check whether the callback exists instead, since that's the only thing we actually need to execute this.

Here and below.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:149]: This method is called "check_method", but it's also sending start. The code would be easier to follow if this function simply returned true/false, and `prov_capabilities` checked this before calling `send_start`/`prov_failed`. `send_start` can use `bt_mesh_prov_link.oob_method` and similar parameters as input.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:703]:
```suggestion
	prov_set_method(AUTH_METHOD_STATIC, 0, 0);
```
so we don't end up sending non-zero by accident after refactoring.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:149]: I thought about it, but I didn't.

File: subsys/bluetooth/mesh/provisioner.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:185]: Why not check the output action? If Device only support input number, and provisioner only support output string, this will be error.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/provisioner.c]: because the oob actions in the prov struct represent the authentication methods that are supported on the device in the provisionee role, just like the static oob value is the local static oob value. Whether this device supports blinking, pushing and so on isn't really relevant when acting in the provisioner role

Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c]: Okay,The specification indeed doesn't require how provisioner to input and output.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:219]:
```suggestion
			BT_WARN("Not support input");
```

Here and above.

File: subsys/bluetooth/mesh/provisioner.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/provisioner.c:278]:
```suggestion
	send_start();
```

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:132]: This sentence is no longer true.
```suggestion
	 *  When this callback returns, the provisioner will start 
	 *  authentication with the chosen method.
```

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Review by trond-snekvik — APPROVED** [review]

File: include/bluetooth/mesh/main.h
Comment (jhedberg) [inline include/bluetooth/mesh/main.h:136]: I realise this has already been happening with other `bt_mesh_prov` struct members, but it surprises me that there are no preprocessor conditionals here to include some members only if provisioner support has been enabled.

File: include/bluetooth/mesh/main.h
Comment (jhedberg) [inline include/bluetooth/mesh/main.h:67]: The comment above the struct name and the struct name seem to be in disagreement here. Shouldn't it (based on the comment) be `struct bt_mesh_dev_capabilities`?

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: include/bluetooth/mesh/main.h
Comment (LingaoM) [reply include/bluetooth/mesh/main.h:136]: you mean change as:

``` C

#if defined(CONFIG_BT_MESH_PROVISIONER)
	void         (*capabilities)(const struct bt_mesh_prov_capabilities *cap);
#endif
```

Comment (jhedberg) [reply include/bluetooth/mesh/main.h:136]: Yes, except it'd probably make sense to group all of them so that you have a single `#if - #endif` pair. Anyway, as I said, this is not new, so this kind of stuff (if it makes sense) can be done in a later PR

File: subsys/bluetooth/mesh/Kconfig
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/Kconfig]: We've been using "provisionee" in the Nordic stacks, as it's not ambiguous like "device". I suggested using provisionee in the kconfig entry to align it with the file names, and I think it's the right name for this role, even if it isn't in the spec.

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: I agree about the file name alignment, what I meant was that if we change it in one place then we change it everywhere. Need to think a little bit more about this, since generally I've tried to keep the policy of not introducing terms for features which can't be looked up anywhere in the specs. Maybe it's just because I participated in a lot of the early Mesh WG meetings prior to 1.0 that I got the concept of "device" vs "node" burned into my brain and therefore don't find it ambiguous or confusing at all 😄

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig]: So will I go back again?

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: @LingaoM no, don't do anything for now

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: @LingaoM @trond-snekvik so is the consensus between you two that you don't want to use the terms "device" or "unprovisioned device" anywhere in the Mesh implementation?

File: include/bluetooth/mesh/main.h
Comment (jhedberg) [reply include/bluetooth/mesh/main.h]: Alternatively the provisioner-specific members could be moved out to a separate struct that's registered independently.

File: subsys/bluetooth/mesh/Kconfig
Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: Btw, I just took a look at BlueZ, and there we have `prov-initiator.c` and `prov-acceptor.c` (not saying we should copy it, but just collecting up examples of how these are named in various implementations). What other open source mesh stacks are there, or even closed-source with a public API?

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: > We've been using "provisionee" in the Nordic stacks

@trond-snekvik did you have any native English speakers in deciding this? Robin is usually super pedantic when it comes to usage of English language in the specs.

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: ..meaning I'm surprised he wouldn't have suggested "provisionee" unless there was some concern with it

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig]: IMO, in provision, the use of `provisionee` is more appropriate than `device`.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/Kconfig]: > did you have any native English speakers in deciding this?
I don't think so. It's definitely a made up word, so I wouldn't be surprised if the WG torpedoed it.

In other mesh APIs, it's a mix of every possible word:
- Nordic: [Provisionee](https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.meshsdk.v4.2.0/group__NRF__MESH__PROV.html?cp=7_2_4_10_1)
- Packetcraft: [Provisioning server](https://github.com/packetcraft-inc/stacks/blob/master/ble-mesh-profile/include/mesh_prv_sr_api.h)
- Silabs: [Provisioning device (PDF!)](https://www.silabs.com/documents/public/reference-manuals/bluetooth-le-and-mesh-software-api-reference-manual.pdf)
-  ESP: [Node](https://github.com/espressif/esp-idf/blob/1c7607d4a67ff96fa532cc4f47838f5ce90b834c/components/bt/esp_ble_mesh/mesh_core/prov.c) (this is a port of the Zephyr stack)
- Cypress: [Provisioning server/node](https://cypresssemiconductorco.github.io/btsdk-docs/BT-SDK/20706-A2_Bluetooth/API/group__wiced__bt__mesh__core.html#gacc7311d144020069d287dc02c6a3a8e1)
- ST Android API: [Target device (PDF!)](https://www.st.com/resource/en/user_manual/dm00434788-bluenrg-mesh-android-api-guide-for-mesh-over-bluetooth-low-energy-stmicroelectronics.pdf)
- Dialog: [Device (PDF!)](https://www.dialog-semiconductor.com/sites/default/files/um-b-098_da1468x_mesh_sdk_rev1v3.pdf)

Only Nordic is using "Provisionee" though, so it's definitely not the conventional name :smile: Perhaps `PROV_DEVICE` is a non-ambiguous middle ground?

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h]: +1 on separating them to a different struct, the application typically wouldn't implement both roles, and if they do, I'd expect they'd be handled by two separate modules. This would also allow proper documentation of the new usage of the OOB auth callbacks

Comment (LingaoM) [reply include/bluetooth/mesh/main.h]: This is a missing problem, which is worth solving in the future PR, or @trond-snekvik  can you amend these changes to this PR.

File: subsys/bluetooth/mesh/Kconfig
Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig]: done

Comment (jhedberg) [reply subsys/bluetooth/mesh/Kconfig]: prov_dev or prov_device sounds good to me. @LingaoM you might want to reflect that in the c-filenames as well.

**Comment by LingaoM** [comment]
okay

在 2020年10月23日星期五，Johan Hedberg <notifications@github.com> 写道：

File: subsys/bluetooth/mesh/Kconfig
Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig]: Done.

File: include/bluetooth/mesh/main.h
Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h]: Let's do it in a separate PR

**Comment by LingaoM** [comment]
Are there any other questions before the code is merged? @jhedberg

**Comment by jhedberg** [comment]
> Are there any other questions before the code is merged? @jhedberg

@LingaoM GitHub will not let me merge this as long as the build is failing. Please fix that first.

**Comment by LingaoM** [comment]
> > Are there any other questions before the code is merged? @jhedberg
> 
> @LingaoM GitHub will not let me merge this as long as the build is failing. Please fix that first.

Done. All check passed.

File: subsys/bluetooth/mesh/Kconfig
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/Kconfig:25]:
```suggestion
	bool "Provisioning device role support"
```

**Review by trond-snekvik — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Comment by LingaoM** [comment]
@tsvehagen Do you have any comments?

**Comment by tsvehagen** [comment]
> @tsvehagen Do you have any comments?

LGTM :+1:

**Review by tsvehagen — APPROVED** [review]

**Pull request merged**
